core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/DcInferringLoadBalancingPolicy.java

```diff
@@ -0,0 +1,77 @@
+/*
```

Comment:
Do we really need an explicit type hierarchy for the different policies here?  The only thing the subtypes do is provide a different helper impl (which is used in a single function at initialization time within the policy).  Seems like the best answer is to just pass the helper as an arg to init(), but I assume we can't do that due to API breakage.  If that's the case can we just pass the helper as an arg to the constructor?  Are we not doing this because it... would also change the contract, this time with the configuration?

Reply:
> The only thing the subtypes do is provide a different helper impl (which is used in a single function at initialization time within the policy)

Not exactly, the type hierarchy is also used to modify the query plan computation logic from OSS (simple replica selection) to DSE (busy node avoidance). And at least in theory, the node filter logic could be customized too. 

> Seems like the best answer is to just pass the helper as an arg to init()

I avoided that because it requires creating a secondary constructor that accepts the helpers as arguments. I'm not sure how well this plays with the reflection mechanism in the driver that instantiates the components by reflection.

Reply:
You are however right about something: it would be a bit better to hide the helper classes from the internal API in the `BasicLoadBalancingPolicy` class hierarchy (concretely speaking, remove references to the helpers from protected methods). I can do that quickly.

Reply:
Sorry, I should've been clearer.  The only thing these subtypes do is offer a different helper implementation.  Other subtypes are free to modify other behaviours as they wish.  And in this case the DSE impl isn't actually a problem since it's not part of this type hierarchy; it just implements the LBP interface directly does it not?

Regardless, I definitely like the idea of removing the helper classes as an explicit part of the API; that addresses a large part of my concerns above.  But I can't help but wonder... now that we don't actually expose the helper classes why have them as separate classes at all?  Why not just make, say, InferringLocalDcHelper.discoverLocalDc() be the impl of DcInferringLoadBalancingPolicy.discoverLocalDc()?

Reply:
> And in this case the DSE impl isn't actually a problem since it's not part of this type hierarchy; it just implements the LBP interface directly does it not?

No, the DSE LBPs both inherit from `DefaultLoadBalancingPolicy`.

> Why not just make, say, InferringLocalDcHelper.discoverLocalDc() be the impl of DcInferringLoadBalancingPolicy.discoverLocalDc()

Because we also have a `DseDcInferringLoadBalancingPolicy` that inherits its query plan logic from `DseLoadBalancingPolicy`, but exhibits a dc inference logic equal to the one exposed by `DcInferringLoadBalancingPolicy`, which is not a parent superclass. If the dc inference logic was baked into `DcInferringLoadBalancingPolicy`, it would have to be copied into `DseDcInferringLoadBalancingPolicy`.